File: app/src/main/java/ru/prsolution/winstrike/mvp/presenters/MainScreenPresenter.java
package ru.prsolution.winstrike.mvp.presenters;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;

import java.util.ArrayList;
import java.util.List;

import ru.prsolution.winstrike.R;
import ru.prsolution.winstrike.common.utils.Utils;
import ru.prsolution.winstrike.mvp.apimodels.Order;
import ru.prsolution.winstrike.mvp.apimodels.OrderModel;
import ru.prsolution.winstrike.mvp.apimodels.Orders;
import ru.prsolution.winstrike.mvp.models.FCMModel;
import ru.prsolution.winstrike.mvp.models.MessageResponse;
import ru.prsolution.winstrike.mvp.models.ProfileModel;
import ru.prsolution.winstrike.mvp.views.MainScreenView;
import ru.prsolution.winstrike.networking.NetworkError;
import ru.prsolution.winstrike.networking.Service;
import ru.prsolution.winstrike.ui.Screens;
import ru.terrakok.cicerone.Router;
import rx.Subscription;
import rx.subscriptions.CompositeSubscription;
import timber.log.Timber;

import static ru.prsolution.winstrike.common.utils.Utils.formatTime;

/**
 * Created by terrakok 25.11.16
 */
@InjectViewState
public class MainScreenPresenter extends MvpPresenter<MainScreenView> {
    private Router router;
    private final Service service;
    private CompositeSubscription subscriptions;

    public MainScreenPresenter(Service service, Router router) {
        this.router = router;
        this.service = service;
        this.subscriptions = new CompositeSubscription();
    }

    public void onCreate() { getViewState().highlightTab(MainScreenView.HOME_TAB_POSITION);
        router.replaceScreen(Screens.START_SCREEN);
    }


    public void onTabPlaceClick(List<OrderModel> mPayList) {
        getViewState().highlightTab(MainScreenView.PLACE_TAB_POSITION);
        router.replaceScreen(Screens.PLACE_SCREEN,mPayList);
    }

    public void onTabUserClick() {
        getViewState().highlightTab(MainScreenView.USER_TAB_POSITION);
        router.replaceScreen(Screens.USER_SCREEN);
    }

    public void onTabHomeClick() {
        getViewState().highlightTab(MainScreenView.HOME_TAB_POSITION);
        router.replaceScreen(Screens.START_SCREEN);
    }

    public void onMapShowClick() {
        router.replaceScreen(Screens.MAP_SCREEN,0);
    }

    public void onChooseScreenClick() {
        getViewState().hideBottomTab();
        router.replaceScreen(Screens.CHOOSE_SCREEN);
    }


    public void onBackPressed() {
//        router.exit(); // used Back command
        getViewState().showBottomTab();
        getViewState().setProfileScreenInterfaceVisibility(false);
        router.replaceScreen(Screens.START_SCREEN,0);
    }


    public void updateProfile(String token, ProfileModel profile, String publicId) {

        Subscription subscription = service.updateUser(new Service.ProfileCallback() {
            @Override
            public void onSuccess(MessageResponse authResponse) {
               getViewState().onProfileUpdateSuccessfully(authResponse);
            }

            @Override
            public void onError(NetworkError networkError) {
                getViewState().onFailtureUpdateProfile(networkError.getAppErrorMessage());
            }

        },token, profile, publicId);

        subscriptions.add(subscription);
    }


    public void sendFCMTokenToServer(String token, FCMModel fcmToken) {

        Subscription subscription = service.sendToken(new Service.FcmTokenCallback() {
            @Override
            public void onSuccess(MessageResponse messageResponse) {
                onTokenSendSuccessfully(messageResponse);
            }

            @Override
            public void onError(NetworkError networkError) {
                onFailtureTokenSend(networkError.getAppErrorMessage());
            }
        }, token, fcmToken);

        subscriptions.add(subscription);
    }


    private void onFailtureTokenSend(String appErrorMessage) {
        Timber.d("On failure send token: %s", appErrorMessage);
    }

    private void onTokenSendSuccessfully(MessageResponse authResponse) {
        Timber.d("Successfully send token: %s", authResponse.getMessage());
    }


    public void getOrders(String token) {
        getViewState().showWait();

        Subscription subscription = service.getOrders(new Service.OrdersCallback() {
            @Override
            public void onSuccess(Orders orders) {
                getViewState().removeWait();

                ArrayList<OrderModel> orderModels = new ArrayList<>();

                for (Order order : orders.getOrders()) {
                    OrderModel orderModel = new OrderModel();
                    orderModel.setCost(order.getCost());
                    orderModel.setAccessCode(order.getAccessCode());
                    orderModel.setStartAt(order.getStartAt());
                    orderModel.setEndAt(order.getEndAt());
                    orderModel.setPcName(order.getPlace().getComputer().getName());

                    String date = Utils.getFormattedDate(order.getStartAt());
                    String time = formatTime(order.getStartAt()) + "-" + formatTime(order.getEndAt());
                    orderModel.setDate(date);
                    orderModel.setTime(time);
                    orderModel.setThumbnail(R.drawable.main_screen);
                    orderModels.add(orderModel);
                }

                getViewState().onGetOrdersSuccess(orderModels);
            }

            @Override
            public void onError(NetworkError networkError) {
                getViewState().removeWait();
                getViewState().onGetOrdersFailure(networkError.getAppErrorMessage());
            }

        },token);

        subscriptions.add(subscription);
    }



    public void onStop() {
        subscriptions.unsubscribe();
    }
}

File: app/src/main/kotlin/ru/prsolution/winstrike/ui/main/MainScreenPresenter.kt
package ru.prsolution.winstrike.ui.main

import com.arellomobile.mvp.MvpPresenter
import ru.prsolution.winstrike.mvp.apimodels.OrderModel
import ru.prsolution.winstrike.mvp.models.FCMModel
import ru.prsolution.winstrike.mvp.models.MessageResponse
import ru.prsolution.winstrike.mvp.models.ProfileModel
import ru.prsolution.winstrike.mvp.views.MainScreenView
import ru.prsolution.winstrike.networking.RetrofitService
import rx.subscriptions.CompositeSubscription
import timber.log.Timber

/**
 * Created by terrakok 25.11.16
 */
class MainScreenPresenter(//    private Router router;
        private val service: RetrofitService) : MvpPresenter<MainScreenView>() {
    private val subscriptions: CompositeSubscription

    init {
        this.subscriptions = CompositeSubscription()
    }//        this.router = router;

    fun onCreate() {
//        viewState.highlightTab(MainScreenView.HOME_TAB_POSITION)
        //        router.replaceScreen(Screens.START_SCREEN);
    }


    fun onTabPlaceClick(mPayList: List<OrderModel>) {
//        viewState.highlightTab(MainScreenView.PLACE_TAB_POSITION)
        //        router.replaceScreen(Screens.PLACE_SCREEN,mPayList);
    }


    fun onTabUserClick() {
//        viewState.highlightTab(MainScreenView.USER_TAB_POSITION)
        //        router.replaceScreen(Screens.USER_SCREEN);
    }

    fun onTabHomeClick() {
//        viewState.highlightTab(MainScreenView.HOME_TAB_POSITION)
        //        router.replaceScreen(Screens.START_SCREEN);
    }

    fun onMapShowClick() {
        //        router.replaceScreen(Screens.MAP_SCREEN,0);
    }

    fun onChooseScreenClick() {
//        viewState.hideBottomTab()
        //        router.replaceScreen(Screens.CHOOSE_SCREEN);
    }


    fun onBackPressed() {
        //        router.exit(); // used Back command
//        viewState.showBottomTab()
//        viewState.setProfileScreenVisibility(false)
        //        router.replaceScreen(Screens.START_SCREEN,0);
    }


    fun updateProfile(token: String, profile: ProfileModel, publicId: String) {

/*        val subscription = service.updateUser(object : Service.ProfileCallback {
            override fun onSuccess(authResponse: MessageResponse) {
                viewState.onProfileUpdateSuccessfully(authResponse)
            }

            override fun onError(networkError: NetworkError) {
                viewState.onFailureUpdateProfile(networkError.appErrorMessage)
            }

        }, token, profile, publicId)

        subscriptions.add(subscription)*/
    }


    fun sendFCMTokenToServer(token: String, fcmToken: FCMModel) {

/*        val subscription = service.sendToken(object : Service.FcmTokenCallback {
            override fun onSuccess(messageResponse: MessageResponse) {
                onTokenSendSuccessfully(messageResponse)
            }

            override fun onError(networkError: NetworkError) {
                onFailtureTokenSend(networkError.appErrorMessage)
            }
        }, token, fcmToken)

        subscriptions.add(subscription)*/
    }


    private fun onFailtureTokenSend(appErrorMessage: String) {
        Timber.d("On failure send token: %s", appErrorMessage)
    }

    private fun onTokenSendSuccessfully(authResponse: MessageResponse) {
        Timber.d("Successfully send token: %s", authResponse.message)
    }


    fun getOrders(token: String) {
//        viewState.showWait()

/*        val subscription = service.getOrders(object : Service.OrdersCallback {
            override fun onSuccess(orders: Orders) {
                viewState.removeWait()

                val orderModels = ArrayList<OrderModel>()

                for (order in orders.orders) {
                    val orderModel = OrderModel()
                    orderModel.arenaName = order.roomName
                    orderModel.cost = order.cost
                    orderModel.accessCode = order.accessCode
                    orderModel.startAt = order.startAt
                    orderModel.endAt = order.endAt
                    orderModel.pcName = order.place.computer.name

                    val date = Utils.getFormattedDate(order.startAt)
                    val time = formatTime(order.startAt) + "-" + formatTime(order.endAt)
                    orderModel.date = date
                    orderModel.time = time
                    orderModel.thumbnail = R.drawable.main_screen
                    orderModels.add(orderModel)
                }

                viewState.onGetOrdersSuccess(orderModels)
            }

            override fun onError(networkError: NetworkError) {
                viewState.removeWait()
                viewState.onGetOrdersFailure(networkError.appErrorMessage)
            }

        }, token)

        subscriptions.add(subscription)*/
    }


    fun onStop() {
        subscriptions.unsubscribe()
    }
}
